Desktop: Fixes #4877: Incorrect list renumbering

austindoupnik opened the pull request:
<!--

Please prefix the title with the platform you are targetting:

Here are some examples of good titles:

- Desktop: Resolves #123: Added new setting to change font
- Mobile, Desktop: Fixes #456: Fixed config screen error
- All: Resolves #777: Made synchronisation faster

And here's an explanation of the title format:

- "Desktop" for the Windows/macOS/Linux app (Electron app)
- "Mobile" for the mobile app (or "Android" / "iOS" if the pull request only applies to one of the mobile platforms)
- "CLI" for the CLI app

If it's two platforms, separate them with commas - "Desktop, Mobile" or if it's for all platforms, prefix with "All".

If it's not related to any platform (such as a translation, change to the documentation, etc.), simply don't add a platform.

Then please append the issue that you've addressed or fixed. Use "Resolves #123" for new features or improvements and "Fixes #123" for bug fixes.

AND PLEASE READ THE GUIDE: https://github.com/laurent22/joplin/blob/dev/CONTRIBUTING.md

-->

These changes are related to/have a minor conflict with PR #4832. Please let me know how I can help get that PR merged, then I can rebase and add some additional test cases. I would also be happy to port the testing changes over to my branch and add test cases.

Here is what I found through manual testing (the expected behavior is the behavior after this patch):

_Note_: **Case C** and **Case D** were working and continue to work, just wanted to test all the code paths.

## Case A

### Steps to Reproduce
1. Create new note
1. Enter text
    ```
    1. item1
    2. item2
    3. item3
    ```
1. Drag select all 3 lines
1. Press numbered list

### Actual
```
4. 1. item1
5. 2. item2
6. 3. item3
```

### Expected
```
item1
item2
item3
```

## Case B

### Steps to Reproduce
1. Create new note
1. Enter text
    ```
    1. item1
    2. item2
    3. item3
    ```
1. Ctrl+A
1. Press numbered list

### Actual
```
2. 1. item1
item2
3. 3. item3
```

### Expected
```
item1
item2
item3
```

## Case C

### Steps to Reproduce
1. Create new note
1. Enter text
    ```
    ```
1. Press numbered list

### Actual
```
1. List item
```

### Expected
```
1. List item
```

## Case D

### Steps to Reproduce
1. Create new note
1. Enter text
    ```
    foobar
    ```
1. Place cursor at the end of the line
1. Press numbered list

### Actual
```
foobar
1. 
```

### Expected
```
foobar
1. 
```

## Case E

### Steps to Reproduce
1. Create new note
1. Enter text
    ```
    1. item1
    1. item2
    1. item3
    ```
1. Drag select all lines or Ctrl+A
1. Press numbered list

### Actual
```
2. 1. item1
3. 1. item2
4. 1. item3
```

### Expected
```
1. item1
1. item2
1. item3
```

laurent22 commented:
The pull request should indeed be based on #4832 once it has been merged. That other pull request also adds test units for this part of the code, so you'll be able to add yours too for the manual tests you mentioned above.

austindoupnik commented:
@laurent22 sounds good, I will rebase and wait for that PR to be merged. It seems it is close. Thank you!

laurent22 commented:
For information the other pull request has been merged now.

austindoupnik commented:
@laurent22 thanks! I updated the PR and ran into one more test case where the behavior is unclear to me what is desired. I also think fixing it requires touching more code than what is covered by `modifyListLines`. The case is somewhat captured in [this test case](https://github.com/laurent22/joplin/pull/4914/files#diff-4167668e44081c1f7608427a8e536d916a6f336bdbf7e88743377de99a1444f4R69).

The following
```
1. item1
item2
2. item3
```

becomes

```
item1
2. item2
item3
```
or
```
item1
3. item2
item3
```

Depending on if it is selected with `Ctrl+A` or click and drag. Testing it quickly with the GH MD editor it seems that GH would do

```
1. 1. item1
2. item2
3. 2. item3
```

Let me know how I should handle this case, and if it should be this PR or another.

laurent22 commented:
Your tests aren't passing at the moment: https://travis-ci.org/github/laurent22/joplin/jobs/769582849#L912

austindoupnik commented:
@laurent22 sorry about that! Haven't quite worked out the workflow yet, forgot to recompile typescript before running the tests.

Updated the test cases. I also removed the `JSON.stringify` I wasn't sure the purpose of that and it made the diffs on assertion failure harder to read as it was diffing JSON instead of the objects directly. I can add it back if it has some purpose. Thanks!

austindoupnik commented on packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useCursorUtils.test.ts:
```diff
@@ -2,19 +2,64 @@ import { modifyListLines } from './useCursorUtils';
+	});
+
+	test('should remove "n. " from each line that has it, and add it to the lines which do not', () => {
+		expect(modifyListLines([...listWithSomeNumbers], 2, '2. ')).toStrictEqual(listWithSomeNumbersAlternated);
```
Here the `2` being passed for `num` can also be `3` depending on how the select block is created. Either with Ctrl+A or by click and drag. That is probably a separate bug that can be fixed as well?

laurent22 replied:
Yes that's probably separate.

lkiThakur commented:
> I also removed the `JSON.stringify` I wasn't sure the purpose of

That was just for checking deep equality; using `toStrictEqual` as you did is better imo.

austindoupnik commented:
@lkiThakur ah thanks for the clarification! I thought it might be something like that

laurent22 commented on packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useCursorUtils.test.ts:
```diff
@@ -2,19 +2,64 @@ import { modifyListLines } from './useCursorUtils';
+		expect(modifyListLines([...listWithOnes], 2, '1. ')).toStrictEqual(listWithNoPrefixes);
+	});
+
+	test('should remove "n. " from each line that has it, and add it to the lines which do not', () => {
```
I don't think this is the best behaviour because if the user selects a block, they want either all the numbers to be gone, or add numbers to each lines, but I don't think they want a mix of both. Perhaps you could check the first item - if it's a number, remove all numbers, and if there's no number, add a number to each line.

austindoupnik replied:
Sure that makes sense. Should it be just the first line or any line?

laurent22 replied:
First line I guess?

laurent22 reviewed (commented):
Thanks for the update, I think it's nearly there. Also for some reason I see that there was no tests added to handle lists with gaps in it. Could you add one please, just to make sure we're preserving this behaviour?

So this:

```
one
two

three
```

should be come this:

```
1.  one
2. two
3.
4. three
```

lkiThakur commented:
> Also for some reason I see that there was no tests added to handle lists with gaps in it.

I wanted to minimise the efforts if switching to other discussed behaviours in future, however if we are preserving this behaviour then someone could add these lines to the code(in the callback of describe)

```javascript
const numberedListWithEmptyLines=[
	
	'1. item1',
	'2. item2',
'3. ',
	'4. item3',
];
const noPrefixListWithEmptyLines = [
'item1',
'item2',
''	   ,
'item3',
];

	test('should add numbers to each line including empty one', () => {
		expect(modifyListLines(noPrefixListWithEmptyLines,1,'1. ')).toEqual(numberedListWithEmptyLines)
	})
```

laurent22 commented:
@austindoupnik, are you still ok with completing this pull request? There are a few tweaks to make, but I think it's nearly there.

austindoupnik commented:
@laurent22 yep! I was out for a few days, but will work on the changes now. thanks!

austindoupnik commented:
@laurent22 FYI

laurent22 commented:
Ok looks good I think, so let's merge

austindoupnik commented:
@laurent22 thank you!
